buildDartApplication: init

Misterio77 opened the pull request:
###### Description of changes

This adds a function for easily packaging non-flutter dart apps (serverside, cli, etc).

Most of the heavy lifting (fetchDartDeps) was already done, I simply wrote some hooks to automatically detect which binaries we need to build and install.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

algorithmiker reviewed (approved):
Overall very good, thanks for your contribution.
I have noted a few small nitpicks but nothing serious .

cc @hacker1024 as I know they had plans for this as well

hacker1024 reviewed (changes requested):
Thanks for the contribution @Misterio77! I have suggested some small changes to the documentation.

A few thoughts:
- The name `buildDartPackage` may give the impression that Nix is used to replace pub, similar to how NodeJS and Python dependencies are all packaged manually. We do not do this, which is probably for the best - version constraints would be a nightmare. As far as I understand, this new derivation function is only really useful for applications. Perhaps the name `buildDartApplication` instead?
- I'm not very sold on the hooks system here. Was there a specific reason for it? It seems to discourage granularity: I think any derivation that needs to build a Dart executable should do so in a dedicated derivation, so that modifications to either the Dart derivation or main derivation do not cause rebuilds of the other.
- I would like to see support for the different methods of Dart compilation (AOT snapshot, JIT snapshot, kernel). AOT snapshots will save space if many use the same Dart SDK, and JIT can be more performant for certain use cases. Kernel may be useful if someone needs to use the resultant binary on another device.
- `buildFlutterApplication` does not have a default hash argument. It will fail to evaluate if none is given, but a fake hash will be used if the given hash is blank. This was modeled after `buildNpmPackage`. I don't mind either way, but the Flutter and Dart behavior should probably be kept consistent.
- Shouldn't the documentation for `buildFlutterApplication` be in a separate PR?

Misterio77 commented:
Working on arguments to make different compile targets, wrapping, and optionally overriding the detected entrypoints (e.g. to build JS)

IndeedNotJames commented:
cc @kira-bruneau because of their work in #228931

Misterio77 commented:
Okay! Pretty cool stuff implemented now.

It's possible to change `dartOutputType` to get any sort of dart output (static bin, aot, jit, kernel, js). Outputs are automatically wrapped depending on the type (`aot-snapshot` uses `dartaotruntime`, `jit-snapshot` and `kernel` uses `dart run`, `js` uses `node`), and this is overridable (setting it to `null` disables wrapping).

It's also now possible to specify custom entry points (that way you won't need to patch the pubspec file or override the phases to e.g. change a bin name).

I plan to document these new features, and split the commits a little.

kira-bruneau commented on pkgs/build-support/dart/build-dart-package/default.nix:
```diff
@@ -0,0 +1,61 @@
+  # Output type to produce. Can be any kind supported by dart
+  # https://dart.dev/tools/dart-compile#types-of-output
+  # If using jit, you might want to pass some arguments to `dartJitFlags`
+, dartOutputType ? "exe"
```
I have very limited knowledge about dart (all of it comes from packaging dart-sass-embedded), but now that this is supporting AOT builds, should it be the default?

The standalone exe build will embed a copy of dart, which seems a bit redundant in the context of a package manager like Nix.

kira-bruneau replied:
Oh maybe exe would make as a default for `stdenv.hostPlatform.isStatic` though!

Misterio77 replied:
Thanks a lot for the feedback!

I think it makes sense doing `aot` by default, yeah:
```sh
# static
$ du -h /nix/store/vc2b4w8vlkbvahcipa2d8sp7byv9hwhd-dart-sass-1.62.1/bin/dart-sass
5.0M	/nix/store/vc2b4w8vlkbvahcipa2d8sp7byv9hwhd-dart-sass-1.62.1/bin/dart-sass
# aot
$ du -h /nix/store/qyk4nx1zn3aaaf22qay9xhcf7aan8c8h-dart-2.19.6/bin/dartaotruntime /nix/store/8823mp1vjdz58804xl860aczq5wkfh6w-dart-sass-1.62.1/share/bin/dart-sass
5.0M	/nix/store/qyk4nx1zn3aaaf22qay9xhcf7aan8c8h-dart-2.19.6/bin/dartaotruntime
4.1M	/nix/store/8823mp1vjdz58804xl860aczq5wkfh6w-dart-sass-1.62.1/share/bin/dart-sass
```

I like that `isStatic` suggestion a lot, really sweet stuff. I'll get it done in a sec

Misterio77 replied:
Actually, hmm, I think it's better to use static builds until we get a separate `dart-runtime` package. The dart derivation is reaaally big:
```sh
$ du -sh /nix/store/qyk4nx1zn3aaaf22qay9xhcf7aan8c8h-dart-2.19.6/bin/
551M	/nix/store/qyk4nx1zn3aaaf22qay9xhcf7aan8c8h-dart-2.19.6/bin/
```

kira-bruneau reviewed (commented):
Oh neat! Thanks for working on this @Misterio77!! :heart:

kira-bruneau commented on pkgs/development/tools/misc/dart-sass/default.nix:
```diff
@@ -0,0 +1,29 @@
+    hash = "sha256-U6enz8yJcc4Wf8m54eYIAnVg/jsGi247Wy8lp1r1wg4=";
+  };
+
+  pubspecLockFile = ./pubspec.lock;
```
Was going to mention it was weird that dart-sass didn't include a `pubspec.lock`, but it looks like @lelgenio already opened a PR for that! So hopefully it'll be in the next release.

lelgenio replied:
Looks like upstream does not want to include it because dart-sass is also a library and must not impose specific version requirements, makes sense: https://github.com/sass/dart-sass/pull/1948#issuecomment-1539185332

Misterio77 committed:
buildDartApplication: init

This adds a function for easily packaging non-flutter dart apps.

Misterio77 committed:
buildFlutterApplication: add docs

Misterio77 committed:
dart-sass: use buildDartApplication

Misterio77 committed:
buildFlutterApplication: default vendorHash to ""

This makes it more ergonomical to use, and is consistent with
buildDartApplication.

kira-bruneau commented on pkgs/build-support/dart/build-dart-application/hooks/dart-install-hook.sh:
```diff
@@ -0,0 +1,31 @@
+        if [ -n "$dartRuntimeCommand" ]; then
+            install -D "$dest" "$out/share/$dest"
+            install -D /dev/null "$out/$dest"
+cat > "$out/$dest" <<EOF
```
Oh that's a good point! I was also thinking of doing that for https://github.com/NixOS/nixpkgs/pull/228931/files#diff-7f45bff25a681f238dff3bfe6b2cc2d15a4383e5912d8d58475dbe0f9c7e7c33R61-R62, and ended up using echo because I thought `makeWrapper` didn't have a "prefix binary" option.

But just thinking about it again, I was going about it the wrong way. You could just use `makeWrapper` on the runtime command binary, and just pass `$dest` as an argument.

Misterio77 replied:
Yeah that's the problem I had too. I remember I considered using `add-flags` like Kira mentioned, but for some reason didn't. I'll fix that in a sec

Misterio77 replied:
Oh I remember now! It's because I need to wrap `jit` and `kernel` with `dart run` specifically. And makeWrapper does not work with that (that's why `$dartRuntimeCommand` is unquoted).

To use `add-flags`, I'd had to conditionally add them based on `dartRuntimeCommand` having spaces or not, and I think that's kind of a pain. What do you folks think?

kira-bruneau replied:
Oh wait a second though, a shebang using `runtimeShell` should be added though.

Misterio77 replied:
Actually, it seems using just `dart` to run them also works. I'll use `wrapProgram` instead.

kira-bruneau replied:
And `\$@` should also be quoted.

kira-bruneau commented:
@hacker1024 I wasn't quite sure what you meant by:

> I'm not very sold on the hooks system here. Was there a specific reason for it? It seems to discourage granularity: I think any derivation that needs to build a Dart executable should do so in a dedicated derivation, so that modifications to either the Dart derivation or main derivation do not cause rebuilds of the other.

Isn't this helper only designed for building dart executables? There would still be a separate derivation for each executable, but it would just use this builder to help reduce the boilerplate.

kira-bruneau reviewed (approved):
This looks good to me! Just wanted to wait to hear back from @hacker1024 before merging this.

kira-bruneau commented:
@hacker1024 Oooh right, I see what you mean now! We'd still need the setup hooks in the builder, but it's not really _necessary_ to expose them as separate packages (rather then just defining them in `setupHooks` for the builder). I guess I was just a bit confused, because pretty much every builder uses hooks, but don't necessarily expose them as packages you can add to `nativeBuildInputs`.

I don't think it really hurts to have more flexibility though! I agree it would be better to break up a multi-language project that uses dart into separate derivations, but giving names to the hooks isn't really adding any more complexity (and we already do this for a bunch of builders like rust + cargo).

kira-bruneau commented:
I think with that though, this would be good to merge as is!
